File: src/mcp.js
/**
 * MCP server integration via @langchain/mcp-adapters.
 *
 * Converts the user's configured MCP servers into LangChain tools that get
 * passed straight into createDeepAgent. The client is cached and only
 * rebuilt when the enabled server set changes (config hash).
 */
import { MultiServerMCPClient } from "@langchain/mcp-adapters";

let cached = null; // { hash, client, tools }

function toAdapterConfig(servers) {
  const mcpServers = {};
  for (const s of servers) {
    // Only streamable HTTP is supported; skip disabled or legacy stdio entries.
    if (!s.enabled || !s.url) continue;
    // streamable HTTP (falls back to SSE automatically inside the adapter)
    mcpServers[s.name] = {
      transport: "http",
      url: s.url,
      headers: s.headers ?? undefined,
    };
  }
  return mcpServers;
}

export async function getMcpTools(servers) {
  const config = toAdapterConfig(servers);
  const names = Object.keys(config);
  if (names.length === 0) {
    if (cached?.client) await cached.client.close().catch(() => {});
    cached = null;
    return { tools: [], errors: [] };
  }

  const hash = JSON.stringify(config);
  const errors = [];
  if (!cached || cached.hash !== hash) {
    if (cached?.client) await cached.client.close().catch(() => {});
    cached = null;
    const client = new MultiServerMCPClient({
      mcpServers: config,
      throwOnLoadError: false,
      prefixToolNameWithServerName: true,
    });
    let tools = [];
    try {
      tools = await client.getTools();
    } catch (e) {
      errors.push(String(e?.message ?? e));
    }
    cached = { hash, client, tools };
  }

  // Per-tool disable is applied on top of the cached connection, so toggling
  // a tool never forces a reconnect. Adapter tool names are `server__tool`.
  const disabled = new Set(
    servers.filter((s) => s.enabled).flatMap((s) =>
      (s.disabledTools ?? []).map((t) => `${s.name}__${t}`)),
  );
  return { tools: cached.tools.filter((t) => !disabled.has(t.name)), errors };
}

export async function closeMcp() {
  if (cached?.client) await cached.client.close().catch(() => {});
  cached = null;
}

/**
 * Test a single server config with a throwaway client and list everything it
 * exposes: tools, prompts and resources (empty arrays when the server does
 * not advertise the capability).
 * @returns {{ok: boolean, tools?: object[], prompts?: object[], resources?: object[], error?: string}}
 */
export async function testMcpServer(config) {
  const name = config.name || "test";
  const map = toAdapterConfig([{ ...config, name, enabled: true }]);
  const client = new MultiServerMCPClient({
    mcpServers: map,
    throwOnLoadError: true,
    prefixToolNameWithServerName: false,
  });
  try {
    const tools = (await client.getTools()).map((t) => ({
      name: t.name,
      description: t.description ?? "",
      // JSON Schema of the tool's input (already dereferenced by the adapter)
      schema: t.schema && typeof t.schema === "object" ? t.schema : null,
    }));
    const raw = await client.getClient(name);
    const tryList = async (fn) => {
      try { return await fn(); } catch { return []; }
    };
    const prompts = await tryList(async () =>
      ((await raw.listPrompts()).prompts ?? []).map((p) => ({
        name: p.name,
        description: p.description ?? "",
        arguments: (p.arguments ?? []).map((a) => ({
          name: a.name, description: a.description ?? "", required: !!a.required,
        })),
      })));
    const resources = await tryList(async () =>
      ((await raw.listResources()).resources ?? []).map((r) => ({
        uri: r.uri,
        name: r.name ?? "",
        description: r.description ?? "",
        mimeType: r.mimeType ?? "",
      })));
    return { ok: true, tools, prompts, resources };
  } catch (e) {
    return { ok: false, error: String(e?.message ?? e) };
  } finally {
    await client.close().catch(() => {});
  }
}

File: app/mcp.py
"""MCP server integration via langchain-mcp-adapters.

Converts the user's configured MCP servers into LangChain tools that get
passed straight into create_deep_agent. Loaded tools are cached and only
refetched when the enabled server set changes (config hash). The Python
adapter opens a fresh MCP session per tool call, so there is no persistent
client to close.

Tool names are prefixed `server__tool` (matching the JS version and the
`disabledTools` convention stored in the app db).
"""
import json

import httpx
from langchain_mcp_adapters.client import MultiServerMCPClient

# 工具列表缓存，按配置哈希失效；仅原地更新，见 get_mcp_tools
_cache: dict = {"hash": None, "tools": {}}


def _httpx_client_factory(headers=None, timeout=None, auth=None) -> httpx.AsyncClient:
    return httpx.AsyncClient(
        headers=headers, timeout=timeout, auth=auth,
        follow_redirects=True, trust_env=False, verify=False,
    )


def _to_adapter_config(servers: list[dict]) -> dict:
    connections = {}
    for s in servers:
        # Only streamable HTTP is supported; skip disabled or legacy stdio entries.
        if not s.get("enabled") or not s.get("url"):
            continue
        connections[s["name"]] = {
            "transport": "streamable_http",
            "url": s["url"],
            "httpx_client_factory": _httpx_client_factory,
            **({"headers": s["headers"]} if s.get("headers") else {}),
        }
    return connections


async def get_mcp_tools(servers: list[dict]) -> tuple[list, list[str]]:
    """Returns (tools, errors). Per-tool disable is applied on top of the
    cached tool list, so toggling a tool never forces a refetch."""
    config = _to_adapter_config(servers)
    if not config:
        _cache.update(hash=None, tools={})
        return [], []

    digest = json.dumps(
        {name: {"url": c["url"], "headers": c.get("headers")} for name, c in config.items()},
        sort_keys=True,
    )
    errors: list[str] = []
    if _cache["hash"] != digest:
        client = MultiServerMCPClient(connections=config)
        by_server: dict[str, list] = {}
        for name in config:
            try:
                tools = await client.get_tools(server_name=name)
            except Exception as e:
                errors.append(f"{name}: {e}")
                continue
            for t in tools:
                t.name = f"{name}__{t.name}"
            by_server[name] = tools
        _cache.update(hash=digest, tools=by_server)

    disabled = {
        f"{s['name']}__{t}"
        for s in servers if s.get("enabled")
        for t in (s.get("disabledTools") or [])
    }
    tools = [
        t
        for server_tools in _cache["tools"].values()
        for t in server_tools
        if t.name not in disabled
    ]
    return tools, errors


async def test_mcp_server(config: dict) -> dict:
    """Test a single server config with a throwaway client and list everything
    it exposes: tools, prompts and resources (empty arrays when the server
    does not advertise the capability)."""
    name = config.get("name") or "test"
    connections = _to_adapter_config([{**config, "name": name, "enabled": True}])
    client = MultiServerMCPClient(connections=connections)
    try:
        tools = [
            {
                "name": t.name,
                "description": t.description or "",
                # JSON Schema of the tool's input (already dereferenced by the adapter)
                "schema": t.tool_call_schema if isinstance(t.tool_call_schema, dict) else (
                    t.args_schema if isinstance(t.args_schema, dict) else None
                ),
            }
            for t in await client.get_tools(server_name=name)
        ]
        prompts: list = []
        resources: list = []
        async with client.session(name) as session:
            try:
                res = await session.list_prompts()
                prompts = [
                    {
                        "name": p.name,
                        "description": p.description or "",
                        "arguments": [
                            {"name": a.name, "description": a.description or "",
                             "required": bool(a.required)}
                            for a in (p.arguments or [])
                        ],
                    }
                    for p in (res.prompts or [])
                ]
            except Exception:
                pass
            try:
                res = await session.list_resources()
                resources = [
                    {"uri": str(r.uri), "name": r.name or "",
                     "description": r.description or "", "mimeType": r.mimeType or ""}
                    for r in (res.resources or [])
                ]
            except Exception:
                pass
        return {"ok": True, "tools": tools, "prompts": prompts, "resources": resources}
    except Exception as e:
        return {"ok": False, "error": str(e)}
